src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -632,12 +633,21 @@ ReactDOMComponent.Mixin = {
           this._currentElement.type
         );
       }
+      var isCustomComponentTag = isCustomComponent(this._tag, props);
+      if (__DEV__ && isCustomComponentTag && !didWarnShadyDom) {
+        warning(
+          !el.shadyRoot,
+          'A component is using shady dom. Using shady dom with React can ' +
+          'cause things to break subtly.'
+        );
+        didWarnShadyDom = true;
```

Comment:
This will set it to true as soon as it encounters the first custom tag, regardless of whether it uses shady DOM.
Maybe make `&& el.shadyRoot` to the `if` and put `didWarnShadyDom` as `warning` argument instead?